modules/core/src/lib/layers/selection-layer.js

```diff
@@ -0,0 +1,162 @@
+// @flow
+/* eslint-env browser */
+import { PolygonLayer, CompositeLayer, COORDINATE_SYSTEM } from 'deck.gl';
+import { point, polygon } from '@turf/helpers';
+import turfBuffer from '@turf/buffer';
+import turfDifference from '@turf/difference';
+
+import { SELECTION_TYPE } from '../deck-renderer/deck-drawer';
+
+import EditableGeoJsonLayer from './editable-geojson-layer';
+
+const defaultProps = {
+  selectionType: SELECTION_TYPE.RECTANGLE,
+  layerIds: [],
+  onSelect: () => {}
+};
+
+const EMPTY_DATA = {
+  type: 'FeatureCollection',
+  features: []
+};
+
+const DEFAULT_SELECTION_INDEX = [0];
+
+const EXPANSION_KM = 50;
+const LAYER_ID_GEOJSON = 'selection-geojson';
+const LAYER_ID_BLOCKER = 'selection-blocker';
+
+export default class SelectionLayer extends CompositeLayer {
+  _selectRectangleObjects(coordinates: any) {
+    const { layerIds, onSelect } = this.props;
+
+    const [x1, y1] = this.context.viewport.project(coordinates[0][0]);
+    const [x2, y2] = this.context.viewport.project(coordinates[0][2]);
+
+    const pickingInfos = this.context.layerManager.pickObjects({
+      viewports: [this.context.viewport],
+      x: Math.min(x1, x2),
+      y: Math.min(y1, y2),
+      width: Math.abs(x2 - x1),
+      height: Math.abs(y2 - y1),
+      layerIds
+    });
+
+    onSelect({ pickingInfos });
+  }
+
+  _selectPolygonObjects(coordinates: any) {
+    const { layerIds, onSelect } = this.props;
+    const mousePoints = coordinates[0].map(c => this.context.viewport.project(c));
+
+    const allX = mousePoints.map(mousePoint => mousePoint[0]);
+    const allY = mousePoints.map(mousePoint => mousePoint[1]);
+    const x = Math.min(...allX);
+    const y = Math.min(...allY);
+    const maxX = Math.max(...allX);
+    const maxY = Math.max(...allY);
+
+    // Use a polygon to hide the outside, because pickObjects()
+    // does not support polygons
+    const landPointsPoly = polygon(coordinates);
+    const bigBuffer = turfBuffer(point(coordinates[0][0]), EXPANSION_KM);
+    let bigPolygon;
+    try {
+      // turfDifference throws an exception if the polygon
+      // intersects with itself (TODO: check if true in all versions)
+      bigPolygon = turfDifference(bigBuffer, landPointsPoly);
```

Comment:
Found a bug with this hack. Notice it is selecting the small stuff in the middle.
![select-bug](https://user-images.githubusercontent.com/469582/48288437-5d9a8580-e429-11e8-9e5f-def3d86a89f1.gif)

Reply:
I think instead of the workaround you're using, we could just do this:
* Get all candidates by doing `pickObjects` with the selection polygon's bounding box
* Iterate over the candidates and do a turf intersects with the selection polygon to filter the candidates to confirmed.

With this approach, shouldn't need the hard-coded 50 kms or the `setTimeout` hack.

Reply:
ok

Reply:
I think we cannot do this because other layers might not have geojson representation of their objects. For example mesh layer.

Reply:
I can make the distance based on zoom and we can improve this hack later....

Reply:
Doing `turfBuffer(landPointsPoly, EXPANSION_KM);` works